nixos/modules/services/admin/pgadmin4.nix

```diff
@@ -0,0 +1,106 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  pkg = pkgs.pgadmin4;
+  cfg = config.services.pgadmin4;
+
+  base = with types; oneOf [ int bool str ];
+
+  formatAttrset = attr:
+    "{${concatStringsSep "\n" (mapAttrsToList (key: value: "${builtins.toJSON key}: ${formatPyValue value},") attr)}}";
+
+  formatPyValue = value:
+    if builtins.isString value then builtins.toJSON value
+    else if value ? _expr then value._expr
+    else if builtins.isInt value then toString value
+    else if builtins.isBool value then (if value then "True" else "False")
+    else if builtins.isAttrset value then (formatAttrset value)
+    else if builtins.isList value then "[${concatStringsSep "\n" (map formatPyValue value)}]"
+    else throw "Unrecognized type";
+
+  formatPy = attrs:
+    concatStringsSep "\n" (mapAttrsToList (key: value: "${key} = ${formatPyValue value}") attrs);
+
+  pyType = with types; attrsOf (oneOf [ (listOf base) (attrsOf base) base ]);
+in
+{
+  options.services.pgadmin4 = {
+    enable = mkEnableOption "PostgreSQL Admin 4";
+
+    port = mkOption {
+      description = "Port for pgadmin4 to run on";
+      type = types.port;
+      default = 5050;
+    };
+
+    initialEmail = mkOption {
+      description = "Initial email for the pgAdmin account.";
+      type = types.str;
+    };
+
+    initialPassword = mkOption {
+      description = "Initial password for the pgAdmin account.";
+      type = types.str;
+    };
+
+    openFirewall = mkEnableOption "firewall passthrough for pgadmin4";
+
+    settings = mkOption {
+      description = ''
+        Settings for pgadmin4.
+        <link xlink:href="https://www.pgadmin.org/docs/pgadmin4/development/config_py.html">Documentation</link>.
+      '';
+      type = pyType;
+      default= {};
+    };
+  };
+
+  config = mkIf (cfg.enable) {
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) [ cfg.port ];
+
+    services.pgadmin4.settings = {
+      DEFAULT_SERVER_PORT = cfg.port;
+      SERVER_MODE = true;
+    } // (optionalAttrs cfg.openFirewall {
+      DEFAULT_SERVER = mkDefault "::";
+    });
+
+    services.postgresql = {
+
+    };
+
+    systemd.services.pgadmin4 = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      # requires = [ "network.target" "postgresql.service" ];
+      requires = [ "network.target" ];
+      wants = [ "postgresql.service" ]; # might not exist
+
+      path = [ config.services.postgresql.package pkgs.coreutils pkgs.bash ];
+
+      preStart = ''
+        # TODO: rewrite to expect script?
+        (
+          # Email address:
+          echo ${escapeShellArg cfg.initialEmail}
+          # Password:
+          echo ${escapeShellArg cfg.initialPassword}
+          # Retype password:
+          echo ${escapeShellArg cfg.initialPassword}
+        ) | ${pkg}/bin/pgadmin4-setup
```

Comment:
the pgadmin4-setup doesn't get used in pgAdmin Server edition (the default). https://github.com/postgres/pgadmin4#configuring-the-python-environment 

> Whilst it is possible to automatically run setup in desktop mode by running the runtime, that will not work in server mode as the runtime doesn't allow command line interaction with the setup program.

I wouldn't use any password / or Email (even initial ones) for pgAdmin4 and let it do its thing (meaning on startup it detects an empty directory and prompts the user on first login for credentials)

Reply:
Currently it has been prompting on startup aswell. I'd need to recheck.

Reply:
Seems to still prompt, even tho it clearly logs server mode. Anything I missed?

```
Feb 23 16:11:56 pgadmin pgadmin4[424574]: NOTE: Configuring authentication for SERVER mode.
Feb 23 16:11:56 pgadmin pgadmin4[424574]: Enter the email address and password to use for the initial pgAdmin user account:
Feb 23 16:11:56 pgadmin pgadmin4[424574]: Email address: 2022-02-23 16:11:56,518: ERROR        pgadmin:        Exception in database migration.
Feb 23 16:11:56 pgadmin pgadmin4[424574]: NOTE: Configuring authentication for SERVER mode.
Feb 23 16:11:56 pgadmin pgadmin4[424574]: Enter the email address and password to use for the initial pgAdmin user account:
Feb 23 16:11:56 pgadmin pgadmin4[424574]: Email address: 2022-02-23 16:11:56,556: ERROR        pgadmin:        Database migration failed
Feb 23 16:11:56 pgadmin pgadmin4[424574]: 2022-02-23 16:11:56,558: ERROR        pgadmin:        Traceback (most recent call last):
```

Reply:
I wouldn't run setup.py at all. For me it worked  by just going to localhost.5050

Reply:
I wasn't running setup.py in this case, just pgadmin4. And that triggered it.

Reply:
Does it trigger when the user isn't root? I've added a user statement, so it runs under the user pgadmin.